### DISCUSSION TITLE
[TlsInterception] Responding to GET/POST method in plugin

### DISCUSSION THREAD

**Discussion by softhub-software-development** [abhinavsingh/proxy.py#870] [Q&A] [answered]
When I access proxy.py, the browser always sends the request as a `CONNECT` method. There is an example plugin `ModifyPostDataPlugin` which seems to do what I am looking for but I never get it to call the proxy with the `POST` method. That's odd, I suppose I am doing something wrong. Could someone please help me getting on track? I'd like to write a plugin, which reacts on `GET` or `POST` requests but the parsers method always indicates being called by CONNECT.

**Comment by abhinavsingh** [comment, marked as answer]
`CONNECT` method is used to open a tunnel with upstream server and are made for all `https` requests.  You cannot inspect or modify what's happening within this (possibly) encrypted connection.  But you can enable TLS interception, which will then allow `proxy.py` to decrypt the content and plugins to handle them as they wish.

**Reply by softhub-software-development** [reply]
Great answer but could you please elaborate on how to enable TLS interception? Is there an example plugin which uses TLS interception? Thanks.

**Reply by abhinavsingh** [reply]
You can find various details here -- https://github.com/abhinavsingh/proxy.py#tls-interception

NOTE: Kindly try the master branch.  TLS interception is currently broken in develop branch, which will be fixed once https://github.com/abhinavsingh/proxy.py/pull/866 is ready for merge.  You can also try to run directly from this PR branch.

**Reply by softhub-software-development** [reply]
When are you going to merge the changes to master? There are right now a few obstacles too many, to make progress on the development of plugins, at least this is my feeling. Best regards.

**Reply by abhinavsingh** [reply]
You can simply use the master branch or at-best use `v2.3.1` from PyPi which is stable release and everything will work as advertised.

**Reply by softhub-software-development** [reply]
Hi,

sorry for intruding again, but would you mind taking a look at the following output and
tell me what's wrong with my TLS configuration? Please let me know how to fix it. Thank you.

```
C:\Users\cnlr\PycharmProjects\proxy.py>curl -v -x localhost:8899 --cacert ca-cert.pem https://httpbin.org/get
*   Trying ::1...
* TCP_NODELAY set
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 8899 (#0)
* allocate connect buffer!
* Establish HTTP proxy tunnel to httpbin.org:443
> CONNECT httpbin.org:443 HTTP/1.1
> Host: httpbin.org:443
> User-Agent: curl/7.55.1
> Proxy-Connection: Keep-Alive
> 
< HTTP/1.1 200 Connection established
< 
* Proxy replied OK to CONNECT request
* CONNECT phase completed!
* schannel: SSL/TLS connection with httpbin.org port 443 (step 1/3)
* schannel: checking server certificate revocation
* schannel: sending initial handshake data: sending 182 bytes...
* schannel: sent initial handshake data: sent 182 bytes
* schannel: SSL/TLS connection with httpbin.org port 443 (step 2/3)
* schannel: failed to receive handshake, SSL/TLS connection failed
* Closing connection 0
* schannel: shutting down SSL/TLS connection with httpbin.org port 443
* Send failure: Connection was aborted
* schannel: failed to send close msg: Failed sending data to the peer (bytes written: -1)
* schannel: clear security context handle
curl: (35) schannel: failed to receive handshake, SSL/TLS connection failed
```

**Reply by abhinavsingh** [reply]
You can ignore ssl check (using flags) or best to trust CA root certificate on your system, which should resolve all such issues.

**Reply by softhub-software-development** [reply]
What do you mean by "ignore ssl check"? I did install the ca certificate (ca-cert.p12) into mmc but nevertheless I get this error.

**Reply by softhub-software-development** [reply]
I guess I should I see an entry in the .proxy/certificates cache if the certificate works, right? How do I debug it if the cache stays empty?

**Reply by abhinavsingh** [reply]
Make sure `openssl` binary is installed on your system.  To perform a quick check, run `make ca-certificates`.

I highly recommend you to read the `README.md`.  Looks like you haven't ran into it yet https://github.com/abhinavsingh/proxy.py#table-of-contents

**Reply by softhub-software-development** [reply]
I am using the instructions given in the README.md. When I ran "make ca-certificates" I noticed this warning, otherwise no errors: RuntimeWarning: 'proxy.common.pki' found in sys.modules after import of package 'proxy.common', but prior to execution of 'proxy.common.pki'; this may result in unpredictable behaviour

Can I ignore this warning? And is there any log that prints information on misconfiguration or any issues related to certificates?

**Reply by abhinavsingh** [reply]
No I don't think that warning will impact anything.  If you are able to generate certificates, all looks good to me.

**Reply by softhub-software-development** [reply]
I tried with 2.3.1 which seems to work better but not perfect. When I access https://www.google.com the https appears in strike through text indicating that the connection is insecure, which makes sense. In the log output, I see the following message:

"File "C:\Users\cnlr\AppData\Local\Programs\Python\Python39\lib\ssl.py", line 1309, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLError: [SSL: SSLV3_ALERT_CERTIFICATE_UNKNOWN] sslv3 alert certificate unknown (_ssl.c:1122)
OSError when wrapping client"

I am using the same ca-certificates as with 2.4.0rc2 and I now find certificates in the cache at ~/.proxy/certificates indicating that there was an attempt to establish a secure connection to Google. Is there any problem with 2.4.0rc2? You said the development version was broken but to me it seems the release candidate is broken too. Any suggestion how to debug that? When will the back port of the development branch be available?

**Reply by abhinavsingh** [reply]
When using `develop` releases, always try the release from `develop` HEAD.   `v2.4.0rc2` happened a while ago and possibly contained a few broken bits.

1. If using container, use latest tag from GHCR https://github.com/abhinavsingh/proxy.py/pkgs/container/proxy.py
2. For pip install, use TestPyPi.  See https://github.com/abhinavsingh/proxy.py/deployments/activity_log?environment=release-testpypi

All released with only `.devN` at the end are from the `develop` branch.   Don't try `.devN.gXXXX` releases, these are PR releases.

**Reply by softhub-software-development** [reply]
I am still puzzled why I can't get my little "man-in-the-middle" plugin to work on 2.4.0.rcX but I made some progress on 2.3.1. There it seems to work, except for a bottleneck which I think may be caused by a too small memoryview passed to handle_upstream_chunk(self, chunk: memoryview) of about 16k. If the upstream is lengthy, this causes trouble, as my plugin should be used to accelerate a download by providing the file locally instead of pulling it from the remote site (right now, the download is much faster than loading it locally). Can you propose a fix or workaround? Thank you.

**Reply by abhinavsingh** [reply]
What doesn't work in `2.4.0`?  Can you kindly expand

**Reply by abhinavsingh** [reply]
Remember to `queue` not `flush` to clients.   `proxy.py` core will make sure data reaches client.  Your plugin should simply queue the packet.

Share the code with us for more insights.  Thank you!!!

**Reply by softhub-software-development** [reply]
It's basically the following, but this works only for small files. How should one handle large files?

	def handle_client_request(self, request: HttpParser) -> Optional[HttpParser]:
		if request.path.decode() == ‘/the_download.zip’:
			self.handle_request_locally(‘/the_download.zip’)
		return request

	def handle_request_locally(self, path) -> None:
		f = open(path, ‘rb’)
		file_data = f.read()
		file_data_size = len(file_data)
		f.close()
		self.client.queue(memoryview(build_http_response(
			httpStatusCodes.OK,
			reason=b’OK’
			headers={
				b’Content-Type’: b’application/octet-stream’
				b’Content-Length: str(file_data_size).encode(‘ascii’)
			},
			body=bytes_(file_data)
		)))

**Reply by abhinavsingh** [reply]
I'll recommend you to use `okResponse` utility, see https://github.com/abhinavsingh/proxy.py/blob/develop/proxy/http/responses.py#L98

```python
self.client.queue(
    okResponse(
        file_data,
        {b'Content-Type': b'application/octet-stream'},
        conn_close=True,
    )
)
```

1) It will automatically take care of `Content-Length` header generation
2) Further, it will also compress (gzip) your content, when its size exceeds `--min-compression-length` flag value
3) `conn_close` will add a `Connection: close` header.  You may or may not want it and is optional.  Working should not get effected by presence of this flag.

Let me know.

**Reply by abhinavsingh** [reply]
Also, how large is your file?

We do have integration tests to guarantee file downloads work as expected , see https://github.com/abhinavsingh/proxy.py/blob/develop/tests/integration/test_integration.sh#L98-L118 

Though files being downloaded by our integration tests are in the range of 30-70 Mb.

**Reply by softhub-software-development** [reply]
I am using debian-11.xxx.iso which is about 3.8 GB for testing. The browser shows 14 days as an approximate time for download (locally).

**Reply by abhinavsingh** [reply]
> I am using debian-11.xxx.iso which is about 3.8 GB for testing. The browser shows 14 days as an approximate time for download (locally).

Thank you for this detail.  We can add a large file test too.

But I notice another issue in your code.  Please read documentation of https://github.com/abhinavsingh/proxy.py/blob/372a9edb4418b11790373038a5a7b969d8db6940/proxy/http/proxy/plugin.py#L104-L119

Your plugin must return `None` if it doesn't want request to go to upstream server.  But looks like your plugin never does that.  Which in-turn means, request is going to upstream server even after you have queued the data locally.

That seems like a reason why you have a `14 days` ETA.  Browser is likely getting corrupt data?

You must replace your code to return None e.g.

`return self.handle_request_locally(‘/the_download.zip’)` will work just fine.
